Math-Interactions-Overview.md

```diff
@@ -0,0 +1,49 @@
+- `guppy-initialization-service.ts`: Service for initializing a guppy instance. On the creator's dashboard, there could be multiple guppy instances present at any given point. This service manages all of these instances and helps in initializing a new instance and finding the currently active instance.
+- `guppy-configuration-service.ts`: Service for handling the global config that will be applied to all guppy instances. 
+- `math-interaction-service.ts`: The core service for handling most of the math interaction-related functionalities on the frontend. This service handles:
+  - All frontend validations using nerdamer.
+  - Modifying the user input from guppy and transforming it such that it is compatible with nerdamer and our backend parser schema.
+  - Generating error messages for various invalid scenarios.
+
+For the custom on-screen keyboard, the functionality is handled by the component present in `core/templates/components/on-screen-keyboard`. This does not extend the custom OSK provided by guppy since customization on that is quite limited. This contains normal buttons that call the guppy API functions to insert characters onto the currently active guppy instance.
+
+### Backend
+There are some custom objects (e.g. `MathEquation`, `AlgebraicIdentifier`, etc.)  present in `extensions/objects/models/objects.py` that help in type checks for the objects used in these interactions. The validators (e.g. `is_valid_math_expression`, `is_valid_algebraic_expression`, etc.) for these objects are present in `core/schema_utils.py`. 
+
+For backend validations of the data stored for these interactions, we have a custom expression parser in `core/domain/expression_parser.py`. **Note that these need to be in sync with the behavior of nerdamer's validations.**
+
+## FAQs
+- **How is the data stored for these interactions validated?**
+  - There are 2 layers of validation for all three math interactions.
+    - **Frontend validations:** This is a set of validations that happen in real-time as soon as the learner/creator enters an expression into guppy. On any change to the guppy input, the input is passed to validation handlers that internally use nerdamer to run syntactic as well as logical checks on the given input. Any errors caught at this stage are displayed to the user in real-time and the submit button is disabled until they are fixed. One caveat to note here is that the checks and display of errors happen when the user clicks outside the input box and not on single-letter changes. This is intentionally done since the latter is an annoying experience. To achieve this, the validations are run when the currently active guppy instance goes out of focus, implying that the user is done typing and validations can be run. Note that these validations run anywhere a guppy instance is present so they will be run on both creator's and learner's view. **These are handled by the `core/templates/services/math-interaction-service.ts`**
```

Comment:
Perhaps give examples of what "syntactic checks" and "logical checks" are.

Reply:
Added to the glossary.